File: haskell/src/Solutions/Euler182.hs
module Solutions.Euler182 where

unconcealed p q =
    let phi = (p - 1) * (q - 1) 
    in sum [e | e <- [2..phi - 1],
                gcd e phi == 1,
                gcd (e - 1) (p - 1) == 2,
                gcd (e - 1) (q - 1) == 2]

euler182 :: IO String
euler182 = return $ show $ unconcealed 1009 3643

File: rust/src/solutions/euler182.rs
use num::integer::gcd;

fn unconcealed(p: i64, q: i64) -> i64 {
    let phi = (p - 1) * (q - 1);
    (2..phi)
        .filter(|&e| gcd(e, phi) == 1)
        .filter(|&e| gcd(e - 1, p - 1) == 2)
        .filter(|&e| gcd(e - 1, q - 1) == 2)
        .sum()
}

pub fn euler182() -> String {
    unconcealed(1009, 3643).to_string()
}
